### PR TITLE
feat(autoscaling): blockDevices property

### LINKED ISSUE #3524: Autoscale Group: Can't Set block_device_mappings
In cloud formation, we can set the block_device_mappings directly for nodes in an autoscale group, but can't access this property through CDK.

On line 403 of auto-scaling-group.ts, the CfnLaunchConfiguration for the group is created but not exposed as a public member. 

CfnLaunchConfiguration has an exposed option to set block_device_mappings, but it can't be set through AutoscalingGroup. 

Requesting that either we can provide block_device_mappings through the AutoscaleGroup constructor, or expose the CfnLaunchConfiguration for modification.

### LINKED ISSUE #3620: Modify the BlockDeviceMappings in AutoScalingGroup construct
**Note: for support questions, please first reference our [documentation](https://docs.aws.amazon.com/cdk/api/latest), then use [Stackoverflow](https://stackoverflow.com/questions/ask?tags=aws-cdk)**. This repository's issues are intended for feature requests and bug reports.

* **I'm submitting a ...**
  - [ ] :beetle: bug report
  - [x] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/aws/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
*If the current behavior is a :beetle:bug:beetle:: Please provide the steps to reproduce*
The current AutoScalingGroup construct cannot be modified the BlockDeviceMappings of it. 
https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_autoscaling/AutoScalingGroup.html

* **What is the expected behavior (or behavior of feature suggested)?**
Sometimes, we need to add or modify the block devices. As in CfnLaunchConfiguration, there's a parameter "block_device_mappings" can modify it. I submit this feature request to have a parameter in the AutoScalingGroup construct.
https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_autoscaling/CfnLaunchConfiguration.html

* **What is the motivation / use case for changing the behavior or adding this feature?**
To add the other block device on instances which is created by ASG.

* **Please tell us about your environment:**

  - CDK CLI Version: xx.xx.xx
  - Module Version: 1.3.0
  - OS: [ OSX Mojave ]
  - Language: [ Python ]

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. associated pull-request, stackoverflow, gitter, etc)

### PR DESCRIPTION
**Pull request by nmussy**
Fixes #3620
Fixes #3524 

----

Please read the [contribution guidelines](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md) and follow the pull-request checklist.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 1]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+   * The device name exposed to the EC2 instance
+   *
+   * @example '/dev/sdh', 'xvdh'
+   *
+   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html
+   */
+  readonly deviceName: string;
+
+  /**
+   *
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: missing docs

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 2]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+   */
+  readonly volume: BlockDeviceVolume;
+
+  /**
+   * Suppresses the device mapping.
+   * If this property is set to true for the root device,  the instance might fail the Amazon EC2 health check.
+   * Amazon EC2 Auto Scaling launches a replacement instance if the instance fails the health check.
+   *
+   * @default device mapping is left untouched
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: use `-` for the @default value if it's just a description

Comment (eladb) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Does this mean `false`?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 3]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+   */
+  readonly noDevice?: boolean;
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Can we avoid the double-negative. Would `mappingEnabled` work (with a `true` default)

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 4]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+}
+
+export interface EbsDeviceOptionsBase {
+  /**
+   * Indicates whether to delete the volume when the instance is terminated.
+   *
+   * @default true for Amazon EC2 Auto Scaling
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: When is it false?

This should be:

```
@default - true for Amazon EC2 auto scaling and false for ...
```

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 5]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+   *
+   * Must only be set for {@link volumeType}: {@link EbsDeviceVolumeType.IO1}
+   *
+   * The maximum ratio of IOPS to volume size (in GiB) is 50:1, so for 5,000 provisioned IOPS,
+   * you need at least 100 GiB storage on the volume.
+   *
+   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html
+   */
+  readonly iops?: number;
+  readonly volumeType?: EbsDeviceVolumeType;
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: missing docs

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 6]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+export interface EbsDeviceOptionsBase {
+  /**
+   * Indicates whether to delete the volume when the instance is terminated.
+   *
+   * @default true for Amazon EC2 Auto Scaling
+   */
+  readonly deleteOnTermination?: boolean;
+
+  /**
+   *
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: ha?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 7]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+  readonly encrypted?: boolean;
+}
+
+export interface EbsDeviceSnapshotOptions extends EbsDeviceOptionsBase {
+  /**
+   * The volume size, in Gibibytes (GiB
+   *
+   * If you specify volumeSize, it must be equal or greater than the size of the snapshot.
+   *
+   * @default snapshot size
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: use `-`

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 8]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+   */
+  public static ephemeral(volumeIndex: number) {
+    if (volumeIndex < 0) {
+      throw new Error(`volumeIndex must be a number starting from 0, got "${volumeIndex}"`);
+    }
+
+    return new this(undefined, `ephemeral${volumeIndex}`);
+  }
+
+  private constructor(public readonly ebsDevice?: EbsDeviceProps, public readonly virtualName?: string) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: make `protected` to allow extensibility by users

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 9]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+ * Describes a block device mapping for an Auto Scaling group.
+ */
+export class BlockDeviceVolume  {
+  /**
+   * Creates a new Elastic Block Storage device
+   *
+   * @param volumeSize The volume size, in Gibibytes (GiB
+   * @param options additional device options
+   */
+  public static ebs(volumeSize: number, options: EbsDeviceOptions = {}): BlockDeviceVolume {
```
Comment (eladb) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: @rix0rrr what do you think? should we use a `from` prefix?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 2 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: It does, but I'll invert it yeah

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 10]
```diff
@@ -864,3 +889,143 @@ export interface MetricTargetTrackingProps extends BaseTargetTrackingProps {
...
+    if (ebsDevice) {
+      const {iops, volumeType} = ebsDevice;
+
+      if (!iops) {
+        if (volumeType === EbsDeviceVolumeType.IO1) {
+          throw new Error('iops property is required with volumeType: EbsDeviceVolumeType.IO1');
+        }
+      } else if (volumeType !== EbsDeviceVolumeType.IO1) {
+        // FIXME won't cause a CloudFormation error, but a good user warning?
+        throw new Error('iops will be ignored without volumeType: EbsDeviceVolumeType.IO1');
```
Comment (nmussy) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Can I remove FIXME and keep the Error?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 2 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 3 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 1 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 4 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 5 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 6 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 7 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 8 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts (hunk 10 above)
Comment (eladb) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: maybe `this.node.addWarning`?

Comment (nmussy) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Done

**Pull request merged**